### PR TITLE
Add Serialization context for codecs and converters

### LINKED ISSUE #1352: Context aware data converters cannot get IDs of outbound activities and child workflow
**Is your feature request related to a problem? Please describe.**
Context aware data converters should have some way to get ID of outbound activities and child workflows when the data converter is being used to serialize activity input or child workflow input. This is useful if the data converter changes behaviour based on the workflow ID, for example using the ID as associated data for encryption.

In the Java SDK this is trivial because the Java SDK will pass the child workflow ID to the data converter when serializing the child workflow input

**Describe the solution you'd like**
I can think of acceptable solutions:
1. Users could create an interceptor to manually set the activity ID/ child workflow ID, maybe the SDK can provide some random generator to help create deterministic UUIDs. Other SDK already have helpers for this so maybe it is worth doing regardless.

2. Ideally the ID would be accessible in the `WithContext` or `WithWorkflowContext` call on the data converter. When we generate the ID in question we could set it on the context and pass that context with the ID to the context aware data converter. This would require some minor refactoring to when the SDK generates an activity ID/ child workflow ID

### PR DESCRIPTION
**Pull request by yuandrew**
A re-creation of https://github.com/temporalio/sdk-go/pull/2149, as that did not properly capture the feature.

## What was changed

  Add SerializationContext — an opt-in mechanism that provides metadata (namespace, workflow ID, activity type,
  etc.) to DataConverter, PayloadCodec, and FailureConverter implementations during serialization/deserialization.

  New public API (converter package):
  - `SerializationContext` sealed interface with two implementations:
    - `WorkflowSerializationContext` — for workflow-level payloads (workflow input/result, child workflow, signal,
  query, update, memo, continue-as-new)
    - `ActivitySerializationContext` — for activity-level payloads (activity input/result, heartbeat, failure details)
  - Three opt-in interfaces: `DataConverterWithSerializationContext`, `PayloadCodecWithSerializationContext`,
  `FailureConverterWithSerializationContext`
  - `WithDataConverterSerializationContext()` / `WithFailureConverterSerializationContext()` helper functions
  - `CodecDataConverter.WithSerializationContext` auto-propagates context to both parent DC and codecs
  - `CompleteActivityWithOptions` / `CompleteActivityByIDWithOptions` / `CompleteActivityByActivityIDWithOptions` / `RecordActivityHeartbeatWithOptions` / `RecordActivityHeartbeatByIDWithOptions` - New client functions that allow users to pass in proper activity metadata to pass into codecs. The existing methods don't have sufficient context information.

<details>

<summary>Context is provided at every serialization site:</summary>

  - Workflow execution (input decode/result encode)
  - Activity and local activity (input encode/decode, result, heartbeat, failure)
  - Child workflow (input encode/decode, result)
  - Signal (including external signal)
  - Query (input/result)
  - Update (input/result/rejection)
  - Continue-as-new (input)
  - Side effect / mutable side effect (uses workflow-scoped DC)
  - Client operations (start, signal, query, terminate, update, describe)
  - Schedule client (create/describe actions)
  - Async activity completion by ID (complete, heartbeat)
  - Workflow task failure and query failure (FC wrapping)

</details>

## Why?

  Enables use cases like per-workflow encryption (using workflow ID as associated data for a key), context-dependent
   encoding, or audit logging — without requiring custom interceptors or thread-local hacks.

  Matches the serialization context feature already implemented in Python, .NET, and Java.

## Checklist

1. Closes #1352 

2. How was this tested:
    - Unit tests in converter/serialization_context_test.go — codec propagation matrix, signing round-trips, helper
  behavior
    - Workflow environment tests in internal/serialization_context_test.go — activity, local activity, child
  workflow, side effect, concurrent operations, signal external, query, update
    - Client mock tests in internal/serialization_context_client_test.go — StartWorkflow, SignalWorkflow,
  QueryWorkflow, TerminateWorkflow
    - Integration tests in test/serialization_context_test.go — end-to-end with signing codec against real server
  (side effect + activity + child workflow combined)

3. Any docs updates needed?

---

> [!NOTE]
> **High Risk**
> High risk because it changes core serialization/encoding behavior across workflows, activities, client APIs, schedules, and the test harness; incorrect context propagation could cause payload decode failures or incompatibilities for users with custom converters/codecs.
> 
> **Overview**
> Adds an opt-in `converter.SerializationContext` API (workflow/activity metadata) and plumbs it through `DataConverter`, `PayloadCodec`, and `FailureConverter` via new `With*SerializationContext` interfaces/helpers and `CodecDataConverter.WithSerializationContext` propagation.
> 
> Updates workflow/activity execution and client paths to wrap converters with the correct workflow/activity context (including pre-generating activity/child workflow IDs so codecs can use them), and stores per-activity/per-child-workflow converters so cancellations/failures decode with matching context.
> 
> Extends the public client API with `*WithOptions` variants for async activity completion/heartbeat that accept activity metadata, and adds broad unit/integration test coverage to validate round-trip behavior and context capture.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c4ec0bf6e06d810b6048f479e40b3993e8b3dc06. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: converter/serialization_context.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+// and may call this method frequently. Avoid recreating expensive objects on every call.
+type FailureConverterWithSerializationContext interface {
+	WithSerializationContext(SerializationContext) FailureConverter
+}
+
+// WithSerializationContext returns a DataConverter that is aware of the given
+// serialization context. If the DataConverter implements
+// [DataConverterWithSerializationContext], it delegates to that implementation;
+// otherwise it returns the original DataConverter unchanged.
+func WithSerializationContext(dc DataConverter, ctx SerializationContext) DataConverter {
```
Comment (jmaeagle99) [inline converter/serialization_context.go]: Maybe rename to `WithDataConverterSerialziationContext` to avoid leading users into thinking this _might_ generically handle all `WithSerializationContext` implementers. I know parameters matter, but type completion doesn't show that many times.

File: converter/serialization_context_test.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+	codec := NewZlibCodec(ZlibCodecOptions{AlwaysEncode: true})
+	cdc := NewCodecDataConverter(parentDC, codec)
+
+	ctx := WorkflowSerializationContext{WorkflowID: "wf-4"}
+	wrapped := WithSerializationContext(cdc, ctx)
+
+	require.Same(cdc, wrapped)
+}
+
+func TestCodecDataConverter_SigningRoundTrip(t *testing.T) {
```
Comment (jmaeagle99) [inline converter/serialization_context_test.go]: nit: Rename to indicate the failure that is being tested.

File: converter/serialization_context_test.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+	require.Equal("hello", result)
+
+	// Different context fails to decode
+	wrapped2 := WithSerializationContext(cdc, WorkflowSerializationContext{WorkflowID: "wf-B"})
+	err = wrapped2.FromPayload(payload, &result)
+	require.Error(err)
+	require.Contains(err.Error(), "signature mismatch")
+}
+
+func TestCodecDataConverter_SigningRoundTrip_ActivityContext(t *testing.T) {
```
Comment (jmaeagle99) [inline converter/serialization_context_test.go]: nit: Rename to indicate the failure that is being tested.

File: converter/serialization_context_test.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+	// Different activity type fails
+	otherCtx := ActivitySerializationContext{WorkflowID: "wf-1", ActivityType: "OtherActivity"}
+	wrapped2 := WithSerializationContext(cdc, otherCtx)
+	err = wrapped2.FromPayload(payload, &result)
+	require.Error(err)
+	require.Contains(err.Error(), "signature mismatch")
+}
+
+func TestWithSerializationContextHelper_NoOp(t *testing.T) {
```
Comment (jmaeagle99) [inline converter/serialization_context_test.go]: `TestDefaultDataConverter_NotSerializationContextAware`?

File: internal/workflow.go [hunk 5]
```diff
@@ -1356,14 +1352,21 @@ func (wc *workflowEnvironmentInterceptor) ExecuteChildWorkflow(ctx Context, chil
...
-		attempt:         1,
+		WorkflowOptions:  *options,
+		Input:            input,
+		WorkflowType:     wfType,
+		Header:           header,
+		scheduledTime:    Now(ctx), /* this is needed for test framework, and is not send to server */
+		attempt:          1,
+		failureConverter: failureConverter,
 	}
+	params.WorkflowID = childWorkflowID
```
Comment (jmaeagle99) [inline internal/workflow.go]: nit: Could this be part of the ExecuteWorkflowParams initialization instead of setting afterwards?

File: internal/internal_event_handlers.go [hunk 6]
```diff
@@ -1495,7 +1494,7 @@ func (weh *workflowExecutionEventHandlerImpl) handleActivityTaskFailed(event *hi
 		&commonpb.ActivityType{Name: activity.activityType.Name},
 		activityID,
 		attributes.GetRetryState(),
-		weh.GetFailureConverter().FailureToError(attributes.GetFailure()),
+		cmp.Or(activity.failureConverter, weh.GetFailureConverter()).FailureToError(attributes.GetFailure()),
```
Comment (jmaeagle99) [inline internal/internal_event_handlers.go]: Under what condition is `activity.failureConverter == nil`? Can we update all the callers to ensure it's not nil and assert that somewhere?

File: converter/serialization_context.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	if sc, ok := dc.(DataConverterWithSerializationContext); ok {
+		return sc.WithSerializationContext(ctx)
```
Comment (jmaeagle99) [inline converter/serialization_context.go]: Should we check that the result is not `nil` and return the original converter instead?

File: converter/serialization_context.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	return dc
+}
+
+// WithFailureConverterSerializationContext returns a FailureConverter that is aware of the given
+// serialization context. If the FailureConverter implements
+// [FailureConverterWithSerializationContext], it delegates to that implementation;
+// otherwise it returns the original FailureConverter unchanged.
+func WithFailureConverterSerializationContext(fc FailureConverter, ctx SerializationContext) FailureConverter {
+	if sc, ok := fc.(FailureConverterWithSerializationContext); ok {
+		return sc.WithSerializationContext(ctx)
```
Comment (jmaeagle99) [inline converter/serialization_context.go]: Same comment about checking result is not `nil`.

File: internal/internal_update.go [hunk 9]
```diff
@@ -170,7 +175,7 @@ func (up *updateProtocol) Reject(err error) {
 			RejectedRequestMessageId:         up.requestMsgID,
 			RejectedRequestSequencingEventId: up.requestSeqID,
 			RejectedRequest:                  up.initialRequest,
-			Failure:                          up.env.GetFailureConverter().ErrorToFailure(err),
+			Failure:                          cmp.Or(up.failureConverter, up.env.GetFailureConverter()).ErrorToFailure(err),
```
Comment (jmaeagle99) [inline internal/internal_update.go]: When is `up.failureConverter == nil`? Can we update all the callers to ensure it's not nil and assert that somewhere?

Comment (jmaeagle99) [reply internal/internal_update.go]: Same for `up.dataConverter`

File: internal/internal_update.go [hunk 10]
```diff
@@ -272,8 +278,13 @@ func defaultUpdateHandler(
 			return
 		}
 
+		if up, ok := callbacks.(*updateProtocol); ok {
+			up.dataConverter = handler.dataConverter
```
Comment (jmaeagle99) [inline internal/internal_update.go:283]: Why is this needed? And should it be the other way around?

```
handler.dataConverter = up.dataConverter
handler.failureConverter = up.failureConverter
```

seems that `up.dataConverter` and `up.failureConverter` are not used beyond this point.

File: internal/internal_workflow_client.go [hunk 11]
```diff
@@ -561,7 +570,10 @@ func (wc *WorkflowClient) RecordActivityHeartbeatByID(ctx context.Context,
 		return err
 	}
 
-	dataConverter := WithContext(ctx, wc.dataConverter)
+	dataConverter := converter.WithSerializationContext(WithContext(ctx, wc.dataConverter), converter.WorkflowSerializationContext{
```
Comment (jmaeagle99) [inline internal/internal_workflow_client.go]: Should this be ActivitySerializationContext?

File: internal/internal_workflow_client.go [hunk 12]
```diff
@@ -497,7 +501,12 @@ func (wc *WorkflowClient) CompleteActivityByID(ctx context.Context, namespace, w
 		return errors.New("empty activity or workflow id or namespace")
 	}
 
-	dataConverter := WithContext(ctx, wc.dataConverter)
+	wfCtx := converter.WorkflowSerializationContext{
```
Comment (jmaeagle99) [inline internal/internal_workflow_client.go]: `ActivitySerializationContext`?

File: internal/internal_workflow_client.go
Comment (jmaeagle99) [inline internal/internal_workflow_client.go]: I think this needs `converter.WithSerializationContext` applied.

File: internal/internal_workflow_client.go (hunk 12 above)
Comment (yuandrew) [reply internal/internal_workflow_client.go]: Added `CompleteActivityByIDWithOptions` to get enough context for `ActivitySerializationContext`

File: internal/internal_workflow_client.go (hunk 11 above)
Comment (yuandrew) [reply internal/internal_workflow_client.go]: added RecordActivityHeartbeatByIDWithOptions to get enough context for `ActivitySerializationContext`

File: internal/internal_update.go (hunk 10 above)
Comment (yuandrew) [reply internal/internal_update.go:283]: These are used in `func (up *updateProtocol) Reject(err error) {` and `func (up *updateProtocol) Complete(success interface{}, outcomeErr error) {`

File: internal/internal_update.go (hunk 9 above)
Comment (yuandrew) [reply internal/internal_update.go]: sure, added default initialization from env in `newUpdateProtocol()`

File: converter/serialization_context.go (hunk 7 above)
Comment (yuandrew) [reply converter/serialization_context.go]: Hmm, if result is `nil`, that is clearly a user bug, and I wanna say in this case, a runtime panic is noisier, thus better than encrypting with the wrong key, or potentially no key at all.

Actually, what do you think about us intentionally panic'ing with something like `panic("WithSerializationContext must not return nil")`, that way users don't see `  panic: runtime error: invalid memory address or nil pointer dereference` and have no idea what's causing it? Changed to this idea, but happy to keep discussing

also added a test to verify this and failureConverter being nil

File: internal/internal_event_handlers.go (hunk 6 above)
Comment (yuandrew) [reply internal/internal_event_handlers.go]: Most places using `cmp.Or` only have their `converter == nil` in test paths, but you bring up a good point, it's probably better to ensure failureConverter is set in tests than have this special code to accept special test scenarios.

File: internal/workflow.go (hunk 5 above)
Comment (yuandrew) [reply internal/workflow.go]: This would go into WorkflowOptions, but yes you're right, it should be fine to set above

File: converter/serialization_context_test.go (hunk 4 above)
Comment (yuandrew) [reply converter/serialization_context_test.go]: good idea

File: converter/serialization_context.go (hunk 1 above)
Comment (yuandrew) [reply converter/serialization_context.go]: good point

File: internal/internal_workflow_client.go [hunk 13]
```diff
@@ -463,28 +467,46 @@ func (wc *WorkflowClient) getWorkflowExecutionHistory(ctx context.Context, rpcMe
...
 	}
 
-	if taskToken == nil {
+	if opts.TaskToken == nil {
 		return errors.New("invalid task token provided")
 	}
 
-	dataConverter := WithContext(ctx, wc.dataConverter)
+	actCtx := converter.ActivitySerializationContext{
+		Namespace:    opts.Namespace,
```
Comment (yuandrew) [inline internal/internal_workflow_client.go]: Chose to intentionally not default back to `wc.Namespace`, and instead have users pass it in, since it would technically be a different namespace

File: internal/error.go [hunk 14]
```diff
@@ -621,7 +621,13 @@ func (wc *workflowEnvironmentInterceptor) NewContinueAsNewError(
 		panic("context is missing required options for continue as new")
 	}
 	env := getWorkflowEnvironment(ctx)
-	workflowType, input, err := getValidatedWorkflowFunction(wfn, args, options.DataConverter, env.GetRegistry())
+	wfInfo := env.WorkflowInfo()
+	wfCtx := converter.WorkflowSerializationContext{
+		Namespace:  wfInfo.Namespace,
+		WorkflowID: wfInfo.WorkflowExecution.ID,
+	}
+	dc := converter.WithDataConverterSerializationContext(getDataConverterFromWorkflowContext(ctx), wfCtx)
```
Comment (jmaeagle99) [inline internal/error.go]: Cursor bot already pointed this out, but it seems that `getDataConverterFromWorkflowContext` already applies the serialization context, so reapplying it again shouldn't be necessary. This seems to be a common pattern throughout this PR.

File: internal/workflow.go [hunk 15]
```diff
@@ -1696,7 +1756,12 @@ func signalExternalWorkflow(ctx Context, workflowID, runID, signalName string, a
 		return future
 	}
 
-	dataConverter := getDataConverterFromWorkflowContext(ctx)
+	dataConverter := converter.WithSerializationContext(
+		getDataConverterFromWorkflowContext(ctx),
+		converter.WorkflowSerializationContext{
+			Namespace:  options.Namespace,
+			WorkflowID: workflowID,
+		})
```
[the remark this answers is not in the document]
Comment (jmaeagle99) [reply internal/workflow.go:1853]: I saw that this was resolved, but it looks like the issue still exists? Or was this somehow fixed somewhere else? Please feel free to resolve again if I missed something.

File: internal/internal_workflow_testsuite.go
Comment (jmaeagle99) [inline internal/internal_workflow_testsuite.go]: This might need to be wrapped with serialization context. The SDK seems to do that, so I think this test infra needs it as well.

File: internal/internal_event_handlers.go [hunk 16]
```diff
@@ -754,18 +755,19 @@ func (wc *workflowEnvironmentImpl) ExecuteActivity(parameters ExecuteActivityPar
 		return ActivityID{}
 	}
 
-	command := wc.commandsHelper.scheduleActivityTask(scheduleID, scheduleTaskAttr, startMetadata)
+	command := wc.commandsHelper.scheduleActivityTask(parameters.ScheduleID, scheduleTaskAttr, startMetadata)
```
Comment (jmaeagle99) [inline internal/internal_event_handlers.go:798]: This looks correct for the Go SDK itself, because `workflowEnvironmentInterceptor.ExecuteActivity` generates the ID. But this affects the `bindings` API, so binding callers now need to call `env.GenerateSequence()` and set `ScheduleID`. As is, I think this _may_ be a breaking change for bindings callers (other SKDs built on top of the Go SDK).

File: internal/serialization_context_client_test.go [hunk 17]
```diff
@@ -0,0 +1,137 @@
+package internal
```
Comment (jmaeagle99) [inline internal/serialization_context_client_test.go:1]: Tests for DescribeWorkflow, PollWorkflowUpdate, and UpdateWithStartWorkflow?

File: internal/serialization_context_test.go [hunk 18]
```diff
@@ -0,0 +1,552 @@
...
+			foundLocal = true
+			s.NotEmpty(actCtx.WorkflowID)
+			s.NotEmpty(actCtx.ActivityType)
+			break
+		}
+	}
+	s.True(foundLocal, "should have captured ActivitySerializationContext with IsLocal=true")
+}
+
+// --- Child Workflow Tests ---
```
Comment (jmaeagle99) [inline internal/serialization_context_test.go]: How about tests for the top-level workflow that has workflow args? I think all of the top-level workflows in these tests do not have args.

File: converter/serialization_context.go [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+	TaskQueue    string
+	IsLocal      bool
+}
+
+func (ActivitySerializationContext) isSerializationContext() {}
+
+// DataConverterWithSerializationContext is an optional interface that [DataConverter]
+// implementations can implement to receive serialization context.
+//
+// When implemented, the SDK calls WithDataConverterSerializationContext before serializing/deserializing
```
Comment (jmaeagle99) [inline converter/serialization_context.go]: nit: While technically correct, I think it's better to say that "the SDK calls WithSerializationContext" since that is the method that `DataConverter` is implementing. The fact that it is indirectly called through `WithDataConverterSerializationContext` shouldn't matter to the developer of the `DataConverter`.

File: internal/internal_workflow_testsuite.go
Comment (yuandrew) [reply internal/internal_workflow_testsuite.go]: good call, added

File: internal/error.go (hunk 14 above)
Comment (yuandrew) [reply internal/error.go]: yeah.. I decided to just remove serialization context from `getDataConverterFromWorkflowContext` and require each caller to explicitly apply context, that way we avoid this double-applying, and it's more clear.

File: internal/internal_event_handlers.go (hunk 16 above)
Comment (yuandrew) [reply internal/internal_event_handlers.go:798]: True, this will need to be called out in the PR description/release notes

File: internal/serialization_context_client_test.go (hunk 17 above)
Comment (yuandrew) [reply internal/serialization_context_client_test.go:1]: added

File: internal/serialization_context_test.go (hunk 18 above)
Comment (yuandrew) [reply internal/serialization_context_test.go]: added

File: internal/internal_event_handlers.go (hunk 16 above)
Comment (yuandrew) [reply internal/internal_event_handlers.go:798]: Ended up adding in backwards compatible checks that will `GenerateSequence()` in the same places we do today if bindings callers don't mirror this new behavior, so we should be good, no longer a breaking change

File: internal/internal_workflow_client.go (hunk 13 above)
Comment (yuandrew) [reply internal/internal_workflow_client.go]: jk, that would be breaking change behavior, will default back to wc.namespace

File: internal/internal_workflow_client.go [hunk 20]
```diff
@@ -497,80 +520,156 @@ func (wc *WorkflowClient) CompleteActivity(ctx context.Context, taskToken []byte
...
+		RunID:      runID,
+		ActivityID: activityID,
+		Result:     result,
+		Err:        err,
+	})
+}
+
+// CompleteActivityByIDWithOptions reports workflow activity completed with full context options.
+func (wc *WorkflowClient) CompleteActivityByIDWithOptions(ctx context.Context, opts CompleteActivityByIDOptions) error {
+	if opts.ActivityID == "" || opts.WorkflowID == "" || opts.Namespace == "" {
```
[the remark this answers is not in the document]
Comment (yuandrew) [reply internal/internal_workflow_client.go:536]: CompleteActivityByID never had the initialization check

File: internal/workflow.go [hunk 21]
```diff
@@ -1034,6 +1036,30 @@ func (wc *workflowEnvironmentInterceptor) ExecuteActivity(ctx Context, typeName
...
 
+	env := getWorkflowEnvironment(ctx)
+	// Generate activity ID before serialization so it's available to context-aware data converters
+	scheduleID := env.GenerateSequence()
+	var activityID string
+	if options.ActivityID != "" {
+		activityID = options.ActivityID
+	} else {
+		activityID = getStringID(scheduleID)
+	}
```
[the remark this answers is not in the document]
Comment (yuandrew) [reply internal/workflow.go:1047]: The old call for `GenerateSequence()` happened at the beginning of `workflowEnvironmentImpl.ExecuteActivity()`. now, they're called in the interceptor immediately before calling `env.ExecuteActivity()`. The order should be the same. In between the current `GenerateSequence` call and where it used to be is building context, wrapping converters, encoding args. There should be no sequence consumption.
